### PR TITLE
Add key cache

### LINKED ISSUE #5183: Journalist API endpoint `get_all_sources` is still too slow
## Description

The `get_all_sources` endpoint of the journalist API is still prohibitively slow with large numbers of sources. Caching source public keys would reduce the time it takes. Also, a cache limit of 1000 for key caching may not be enough for larger installations. Given the relatively small amount of memory these caches require, we should increase that.

### PR DESCRIPTION
**Pull request by rmol**
## Status

Ready for review

## Description of Changes

Moves the CryptoUtil caching to Redis. This reduces total memory usage compared to keeping the caches in the WSGI processes, and eliminates the possibility that the cache would have to be repopulated if WSGI processes restart.

Adds caching of public keys, not just fingerprints.

Renames `CryptoUtil.getkey` and `CryptoUtil.export_pubkey` to better match what they do.

Adds some detail to the output of `create-dev-data.py` to better indicate progress. 

Uses 1024-bit keys if the environment variable `SECUREDROP_ENV` is `test` _or_ `dev`, which speeds up the local dev server startup with large numbers of sources.

Fixes #5183.

## Testing

Run the dev server with the environment variable `NUM_SOURCES` set to a large number, like 500 or 1000. Use the test script from the #5100 test plan. Running it against this branch should show faster response times than master, once the caches are populated.

### Checking Redis memory usage

Stop Apache with `service apache2 stop`.

In `redis-cli`:
- List all keys with `keys *`
- Delete the Redis `CryptoUtil` cache keys if they exist:
  - `del sd/crypto-util/fingerprints`
  - `del sd/crypto-util/keys`
- Execute `info` and look for `used_memory_human`. It should be under
  a megabyte, e.g. `used_memory_human:517.83K`

Start Apache with `service apache2 start`.

Wait for the caches to populate. It took a few minutes on my system. You can watch for the Apache and GPG processes to drop out of `top`, then check the number of keys in the Redis caches in `redis-cli` with `hlen sd/crypto-util/fingerprints` and `hlen sd/crypto-util/keys`.

Execute `info` in `redis-cli` and look for `used_memory_human`. With dev/test 1024-bit keys, it should be under two megabytes, e.g. `used_memory_human:1.66M`. With production 4096-bit keys I see `used_memory_human:2.64M`. 

## Deployment

This requires additional working memory, but as suggested in the test plan, caching 1000 1024-bit keys and fingerprints in Redis takes under 2 megabytes. Production 4096-bit keys will result in a proportional increase.

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

### CODE REVIEW

File: securedrop/create-dev-data.py [hunk 1]
```diff
@@ -58,12 +58,14 @@ def main(staging=False):
 
         # Add test sources and submissions
         num_sources = int(os.getenv('NUM_SOURCES', 2))
-        for i in range(num_sources):
+        for i in range(1, num_sources + 1):
             if i == 0:
```
[the remark this answers is not in the document]
Comment (rmol) [reply securedrop/create-dev-data.py]: excellent catch. fixed.

File: install_files/ansible-base/roles/app/templates/sites-available/journalist.conf [hunk 2]
```diff
@@ -2,8 +2,7 @@ ServerName {{ securedrop_app_apache_listening_address }}
 <VirtualHost {{ securedrop_app_apache_listening_address }}:8080>
 
 WSGIDaemonProcess journalist processes=2 threads=30 display-name=%{GROUP} python-path=/var/www/securedrop
-WSGIProcessGroup journalist
-WSGIScriptAlias / /var/www/journalist.wsgi
+WSGIScriptAlias / /var/www/journalist.wsgi process-group=journalist application-group=journalist
```
Comment (redshiftzero) [inline install_files/ansible-base/roles/app/templates/sites-available/journalist.conf:5]: right now this state is enforced via the Ansible playbooks, and so we would need to migrate this on running instances via postinst in the `securedrop-app-code` package

**Comment by creviera** [comment]
Since [I am still seeing issues with timeouts](https://github.com/freedomofpress/securedrop-workstation/issues/525#issuecomment-610705348), I think it's important to test this PR against a staging server with 200, 300, and 1000 sources if there's time (which would close https://github.com/freedomofpress/securedrop-client/issues/1007 if it succeeds), before it's merged. Also, remember to deleted the local database between runs so you can see whether or not the source list is populated.

**Comment by rmol** [comment]
I've updated the PR description to add detail about the switch to Redis. @redshiftzero this is ready for another look.

@creviera In testing with 1000 sources, I see the client source list populated, with messages being downloaded, in under two minutes. The initial sync completes in around 20 seconds.

File: securedrop/crypto_util.py [hunk 3]
```diff
@@ -254,29 +258,45 @@ def delete_reply_keypair(self, source_filesystem_id):
...
-        fingerprint = self.keycache.get(name)
-        if fingerprint:  # cache hit
+        The supplied name is usually a source filesystem ID.
+        """
+        fingerprint = self.fprcache.get(name)
+        if fingerprint:
             return fingerprint
 
-        # cache miss
+        target_uid = 'Autogenerated Key <{}>'.format(name)
```
[the remark this answers is not in the document]
Comment (rmol) [reply securedrop/crypto_util.py]: I've reverted this change.

File: install_files/ansible-base/roles/app/templates/sites-available/journalist.conf (hunk 2 above)
Comment (rmol) [reply install_files/ansible-base/roles/app/templates/sites-available/journalist.conf:5]: I've added this editing to `postinst`, so it should now happen when the package is updated.

**Comment by rmol** [comment]
The timing in my last comment was observed with the client running in `sd-app`, using the proxy and Tor, talking to my staging VM populated with 1000 sources, each with two submissions and replies, using 1024-bit keys.

**Comment by rmol** [comment]
@kushaldas The cache is [populated](https://github.com/freedomofpress/securedrop/blob/5aca09278a3cfe60ea7865c1d61d1c912aeeef94/securedrop/journalist.py#L22) when Apache starts the [mod_wsgi processes](https://github.com/freedomofpress/securedrop/blob/5aca09278a3cfe60ea7865c1d61d1c912aeeef94/install_files/ansible-base/roles/app/templates/sites-available/journalist.conf#L5) (from the `WSGIScriptAlias` [documentation](https://modwsgi.readthedocs.io/en/latest/configuration-directives/WSGIScriptAlias.html): "If both process-group and application-group options are set, the WSGI script file will be pre-loaded when the process it is to run in is started, rather than being lazily loaded on the first request.")

**Pull request merged**